src/Backups/BackupCoordinationRemote.cpp

```diff
@@ -165,15 +165,28 @@ namespace
     constexpr size_t NUM_ATTEMPTS = 10;
 }
 
-BackupCoordinationRemote::BackupCoordinationRemote(const String & zookeeper_path_, zkutil::GetZooKeeper get_zookeeper_)
+BackupCoordinationRemote::BackupCoordinationRemote(
+    const String & zookeeper_path_, zkutil::GetZooKeeper get_zookeeper_, bool remove_zk_nodes_in_destructor_)
     : zookeeper_path(zookeeper_path_)
     , get_zookeeper(get_zookeeper_)
-    , status_sync(zookeeper_path_ + "/status", get_zookeeper_, &Poco::Logger::get("BackupCoordination"))
+    , remove_zk_nodes_in_destructor(remove_zk_nodes_in_destructor_)
+    , stage_sync(zookeeper_path_ + "/stage", get_zookeeper_, &Poco::Logger::get("BackupCoordination"))
 {
     createRootNodes();
 }
 
-BackupCoordinationRemote::~BackupCoordinationRemote() = default;
+BackupCoordinationRemote::~BackupCoordinationRemote()
+{
+    try
+    {
+        if (remove_zk_nodes_in_destructor)
+            removeAllNodes();
```

Comment:
Seems like race condition is possible if this dtor is called on one host (for example, due to exception) while some hosts are still doing backup.

Reply:
`remove_zk_nodes_in_destructor` is set only for the initiator of a query

Reply:
Race condition is still possible in case like this: Initiator successfully executes `executeDDLQueryOnCluster` (in `doBackup`), but then gets an exception ("session expired", "memory limit exceeded", etc) before all hosts finished doing backup. It unwinds stack in calls this dtor. It removes nodes that may be used by other hosts. 

Another issue is that these nodes are never removed if initiator lost connection. But it's minor issue and maybe we can ignore it for now.

Reply:
> It unwinds stack in calls this dtor. It removes nodes that may be used by other hosts.

Then after trying to write/read to the coordination any information other hosts will stop doing backup because of error `ZNONODE`

Reply:
https://github.com/ClickHouse/ClickHouse/blob/ae3d9fd9623a829bfca4c09bc258e33e0859743f/src/Common/ZooKeeper/ZooKeeper.h#L220-L222

Reply:
Do you mean that "the result is undefined"? Ok, but what that undefined result can be anyway? Either `removeAllNodes()` removes a node created by another host (which it doesn't expect and will throw an exception) or  `removeAllNodes()` removes the parent node for the whole backup operation (which another host doesn't expect again and will throw an exception). The worst thing in both cases that some nodes in ZooKeeper can be never removed after that

Reply:
Yes, I mean that it can partially remove the subtree (remove random subset of nodes) or even do not remove anything (in this case other hosts may continue doing backup).

Reply:
Any backup needs to be finalized is made only by the initiator so other host may do some useless part of work if the initiator fails but we won't get any backup as a result

Reply:
Ok, please add a comment about this and about `ZNONODE`